Firestore/Source/Core/FSTView.mm

```diff
@@ -242,57 +243,73 @@ - (FSTViewDocumentChanges *)computeChangesWithDocuments:(FSTMaybeDocumentDiction
         newDoc = nil;
       }
     }
-    if (newDoc) {
-      newDocumentSet = [newDocumentSet documentSetByAddingDocument:newDoc];
-      if (newDoc.hasLocalMutations) {
-        newMutatedKeys = newMutatedKeys.insert(key);
-      } else {
-        newMutatedKeys = newMutatedKeys.erase(key);
-      }
-    } else {
-      newDocumentSet = [newDocumentSet documentSetByRemovingKey:key];
-      newMutatedKeys = newMutatedKeys.erase(key);
-    }
 
+    BOOL oldDocHadPendingMutations = oldDoc && oldMutatedKeys.contains(oldDoc.key);
+
+    // We only consider committed mutations for documents that were mutated during the lifetime of
+    // the view.
+    BOOL newDocHasPendingMutations =
+        newDoc && (newDoc.hasLocalMutations ||
+                   (oldMutatedKeys.contains(newDoc.key) && newDoc.hasCommittedMutations));
+
+    BOOL changeApplied = NO;
     // Calculate change
     if (oldDoc && newDoc) {
       BOOL docsEqual = [oldDoc.data isEqual:newDoc.data];
-      if (!docsEqual || oldDoc.hasLocalMutations != newDoc.hasLocalMutations) {
-        // only report a change if document actually changed.
-        if (docsEqual) {
-          [changeSet addChange:[FSTDocumentViewChange
-                                   changeWithDocument:newDoc
-                                                 type:FSTDocumentViewChangeTypeMetadata]];
-        } else {
+      if (!docsEqual) {
+        if (![self shouldWaitForSyncedDocument:newDoc oldDocument:oldDoc]) {
           [changeSet addChange:[FSTDocumentViewChange
                                    changeWithDocument:newDoc
                                                  type:FSTDocumentViewChangeTypeModified]];
-        }
+          changeApplied = YES;
 
-        if (lastDocInLimit && self.query.comparator(newDoc, lastDocInLimit) > 0) {
-          // This doc moved from inside the limit to after the limit. That means there may be some
-          // doc in the local cache that's actually less than this one.
-          needsRefill = YES;
+          if (lastDocInLimit && self.query.comparator(newDoc, lastDocInLimit) > 0) {
+            // This doc moved from inside the limit to after the limit. That means there may be
+            // some doc in the local cache that's actually less than this one.
+            needsRefill = YES;
+          }
         }
+      } else if (oldDocHadPendingMutations != newDocHasPendingMutations) {
+        [changeSet
+            addChange:[FSTDocumentViewChange changeWithDocument:newDoc
+                                                           type:FSTDocumentViewChangeTypeMetadata]];
+        changeApplied = true;
```

Comment:
s/true/YES/ ?

If so, also lines 283, 288.
If not, do the reverse on lines 255, 264.